File: js/testStructural.js
import { RealSubject, MyProxy, ProxyFunction } from "./2_1_proxy";
import { ConcreteComponent, ConcreteDecoratorA, ConcreteDecoratorB, DecoratorFunction } from "./2_2_decorator";
import { Adaptee, Adapter, adapt } from "./2_3_adapter";
import CultureFacade from "./2_4_facade";
import { SongSmallWidget, SongMiddleWidget, SongBigWidget, BookSmallWidget, BookMiddleWidget, BookBigWidget, Book, Song } from "./2_5_0_no_bridge";
import { SmallWidgetAbstraction, MiddleWidgetAbstraction, BigWidgetAbstraction, SongWidgetData, BookWidgetData } from "./2_5_1_bridge";
import { SongSmallWidgetWithMixin, SongMiddleWidgetWithMixin, SongBigWidgetWithMixin, BookBigWidgetWithMixin, BookMiddleWidgetWithMixin, BookSmallWidgetWithMixin } from "./2_5_2_mixins";

export default {
    testNoBridge: function () {
        const song = new Song("Вставай!", "Вставай! Пий чай з молоком, Молися на теплий душ!");
        const book = new Book("Шаблони проєктування: Елементи повторно використовуваного об'єктно-орієнтованого програмного забезпечення",
            "книга 1994 року з програмної інженерії, в якій запропоновані і описані архітектурні рішення деяких частих проблем у проєктуванні ПЗ");
        const widgets = [
            new BookSmallWidget(book),
            new BookMiddleWidget(book),
            new BookBigWidget(book),
            new SongSmallWidget(song),
            new SongMiddleWidget(song),
            new SongBigWidget(song)
        ];
        for (let widget of widgets)
            console.log(widget.render());
    },
    testMixin: function(){
        const song = new Song("Вставай!", "Вставай! Пий чай з молоком, Молися на теплий душ!");
        const book = new Book("Шаблони проєктування: Елементи повторно використовуваного об'єктно-орієнтованого програмного забезпечення",
            "книга 1994 року з програмної інженерії, в якій запропоновані і описані архітектурні рішення деяких частих проблем у проєктуванні ПЗ");
        const widgets = [
            new BookSmallWidgetWithMixin(book),
            new BookMiddleWidgetWithMixin(book),
            new BookBigWidgetWithMixin(book),
            new SongSmallWidgetWithMixin(song),
            new SongMiddleWidgetWithMixin(song),
            new SongBigWidgetWithMixin(song)
        ];
        for (let widget of widgets)
            console.log(widget.render());
    },
    testBridge: function () {
        const widgets = [
            new SmallWidgetAbstraction(),
            new MiddleWidgetAbstraction(),
            new BigWidgetAbstraction()
        ];
        const song = new Song("Вставай!", "Вставай! Пий чай з молоком, Молися на теплий душ!");
        const book = new Book("Шаблони проєктування: Елементи повторно використовуваного об'єктно-орієнтованого програмного забезпечення",
            "книга 1994 року з програмної інженерії, в якій запропоновані і описані архітектурні рішення деяких частих проблем у проєктуванні ПЗ");
        const widgetData = [
            new SongWidgetData(song),
            new BookWidgetData(book)
        ];
        for (let widget of widgets)
            for (let data of widgetData)
                console.log(widget.render(data));
    },
    testFacade: function () {
        const music = new CultureFacade("music");
        console.log(music.get(3));
        const movie = new CultureFacade("movie");
        console.log(movie.get(5));
    },
    testAdapter: function () {
        let oldLib = new Adaptee();
        console.log(oldLib.getSpecificRequest("Test", 4, true));
        let newLib = new Adapter(oldLib);
        console.log(newLib.getRequest("Test"));
        console.log("============================");
        let funcLib = adapt(oldLib);
        console.log(funcLib.getRequest("Test"));
    },
    testProxy: function () {
        let subject = new RealSubject();
        console.log(subject.request());
        subject = new MyProxy(subject);
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
        subject = ProxyFunction(subject);
        console.log('===========================');
        subject = ProxyFunction(new RealSubject());
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
        console.log(subject.request());
    },
    testDecorator: function () {
        let component = new ConcreteComponent();
        console.log(component.operation());
        component = new ConcreteDecoratorA(component);
        console.log(component.operation());
        component = new ConcreteDecoratorB(component);
        console.log(component.operation());

        let component2 = DecoratorFunction(new ConcreteComponent());
        console.log(component2.operation());
        console.log(component2.otherOperation());
    }
}

File: cs/TestStructural.cs
using System;
using Structural.Proxy;
using Structural.Decorator;
using Structural.Game;
using Structural.Adapter;
using Structural.Bridge;
using Structural.Flyweight;

namespace Test
{
    class StructuralPaterns
    {
        public static void TestFlyweight()
        {
            // Клиентский код обычно создает кучу предварительно заполненных
            // легковесов на этапе инициализации приложения.
            var factory = new FlyweightFactory(
                new Car { Company = "Chevrolet", Model = "Camaro2018", Color = "pink" },
                new Car { Company = "Mercedes Benz", Model = "C300", Color = "black" },
                new Car { Company = "Mercedes Benz", Model = "C500", Color = "red" },
                new Car { Company = "BMW", Model = "M5", Color = "red" },
                new Car { Company = "BMW", Model = "X6", Color = "white" }
            );
            factory.listFlyweights();

            addCarToPoliceDatabase(factory, new Car
            {
                Number = "CL234IR",
                Owner = "James Doe",
                Company = "BMW",
                Model = "M5",
                Color = "red"
            });

            addCarToPoliceDatabase(factory, new Car
            {
                Number = "CL234IR",
                Owner = "James Doe",
                Company = "BMW",
                Model = "X1",
                Color = "red"
            });

            factory.listFlyweights();
        }

        public static void addCarToPoliceDatabase(FlyweightFactory factory, Car car)
        {
            Console.WriteLine("\nClient: Adding a car to database.");

            var flyweight = factory.GetFlyweight(new Car
            {
                Color = car.Color,
                Model = car.Model,
                Company = car.Company
            });

            // Клиентский код либо сохраняет, либо вычисляет внешнее состояние и
            // передает его методам легковеса.
            flyweight.Operation(car);
        }
        public static void TestBridge()
        {
            Abstraction abstractionA = new Abstraction(new ConcreteImplementationA());
            Abstraction abstractionB = new Abstraction(new ConcreteImplementationB());
            Abstraction exAbstractionA = new ExtendedAbstraction(new ConcreteImplementationA());
            Abstraction exAbstractionB = new ExtendedAbstraction(new ConcreteImplementationB());
            Console.WriteLine(abstractionA.Operation());
            Console.WriteLine(abstractionB.Operation());
            Console.WriteLine(exAbstractionA.Operation());
            Console.WriteLine(exAbstractionB.Operation());
        }
        public static void TestProxy()
        {
            ISubject subject = new RealSubject("8.8.8.8");
            Console.WriteLine(subject.Request());
            subject = new Proxy(subject as RealSubject);
            Console.WriteLine(subject.Request());
            Console.WriteLine(subject.Request());
            subject = new Proxy("127.0.0.1");
            Console.WriteLine(subject.Request());
        }

        public static void TestDecorator()
        {
            IComponent component = new ConcreteComponent();
            Console.WriteLine(component.Operation());
            Decorator decorator = new ConcreteDecoratorA(component);
            Console.WriteLine(decorator.Operation());
            decorator = new ConcreteDecoratorB(decorator);
            Console.WriteLine(decorator.Operation());
        }

        public static void TestGame()
        {
            IDamageActor humen = new Character("Humen", 300, 50);
            IDamageActor orc = new Character("Orc", 350, 75);
            humen = new DefenceBuff(new DefenceBuff(humen, 10), 30);

            while (!humen.IsDead() && !orc.IsDead())
            {
                humen.Hit(orc);
                orc.Hit(humen);
            }
        }

        public static void TestAdapter()
        {
            Adaptee oldLib = new Adaptee();
            Console.WriteLine(oldLib.GetSpecificRequest("Test", 4, true));
            Adapter newLib = new Adapter(oldLib);
            Console.WriteLine(newLib.GetRequest("Test"));
        }

        public static void TestAdapterWinApi()
        {
            Process process = new Process();
            AdaptedPocess adaptedPocess = new AdaptedPocess(process);
            adaptedPocess.Create("notepad.exe");
        }
    }
}